errors/functional_test.go

```diff
@@ -0,0 +1,281 @@
+package errors_test
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+	"testing"
+
+	"github.com/ahmetb/go-linq"
+	"github.com/mailgun/holster/v5/callstack"
+	"github.com/mailgun/holster/v5/errors"
+	"github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+type TestError struct {
+	Msg string
+}
+
+func (e *TestError) Error() string {
+	return e.Msg
+}
+
+func (e *TestError) Is(target error) bool {
+	_, ok := target.(*TestError)
+	return ok
+}
+
+func TestContext(t *testing.T) {
+	// Wrap an error with context
+	err := &TestError{Msg: "query error"}
+	wrap := errors.WithContext{"key1": "value1"}.Wrap(err, "message")
+	assert.NotNil(t, wrap)
+
+	// Unwrap should return TestError
+	u := errors.Unwrap(wrap)
+	require.NotNil(t, u)
+	assert.Equal(t, "query error", u.Error())
+
+	// Extract as normal map
+	m := errors.ToMap(wrap)
+	require.NotNil(t, m)
+	assert.Equal(t, "value1", m["key1"])
+
+	// Can use errors.Is() from std `errors` package
+	assert.True(t, errors.Is(err, &TestError{}))
+	assert.True(t, errors.Is(wrap, &TestError{}))
+
+	// Can use errors.As() from std `errors` package
+	myErr := &TestError{}
+	assert.True(t, errors.As(wrap, &myErr))
+	assert.Equal(t, myErr.Msg, "query error")
+
+	// Extract as Logrus fields
+	f := errors.ToLogrus(wrap)
+	require.NotNil(t, f)
+	b := bytes.Buffer{}
+	logrus.SetOutput(&b)
+	logrus.WithFields(f).Info("test logrus fields")
+	logrus.SetOutput(os.Stdout)
+	assert.Contains(t, b.String(), "test logrus fields")
+	assert.Contains(t, b.String(), `excValue="message: query error"`)
+	assert.Contains(t, b.String(), `excType="*errors_test.TestError"`)
+	assert.Contains(t, b.String(), "key1=value1")
+
+	assert.Equal(t, "message: query error", wrap.Error())
+
+	// Should output the message and fields
+	out := fmt.Sprintf("%+v", wrap)
+	assert.True(t, strings.Contains(out, `message: query error (key1=value1)`))
+}
+
+func TestWithErrorf(t *testing.T) {
+	err := errors.New("this is an error")
+	wrap := errors.WithContext{"key1": "value1", "key2": "value2"}.Wrap(err, "message")
+	err = fmt.Errorf("wrapped: %w", wrap)
+
+	out := fmt.Sprintf("final: %s", err)
+	assert.Contains(t, out, "final: wrapped: message: this is an error")
```

Comment:
`assert.Equals` would provide a better picture.

Reply:
unfortunately, maps are randomized so the key/values will be in different order each test run.